### DISCUSSION TITLE
Monaco 2.0 - Problem with management zone configuration (update instead of create)

### DISCUSSION THREAD

**Discussion by mackermann2** [Dynatrace/dynatrace-configuration-as-code#916] [Q&A] [answered]
Hi everyone, 
I try to use the latest version of monaco to create management zones but I have a problem when I tried to use an environment variable in the configuration file YAML. Indeed, please find the JSON and YAML files bellow : 

**management-zone.yaml**
```
configs:
        - id: create-ManagementZone
          type:
                settings:
                        schema: builtin:management-zones
                        scope: environment
          config:
                  name: 
                        type: environment
                        name: MGT_ZONE_NAME
                  parameters:
                        contractID: 
                                type: environment
                                name: contractID
                  template: management-zone.json

```
**management-zone.json**
```
{
  "name": "{{ .name }}",
  "rules": [
    {
      "enabled": true,
      "type": "ME",
      "attributeRule": {
        "entityType": "ESXI_HOST",
        "conditions": [
          {
            "key": "ESXI_HOST_NAME",
            "operator": "CONTAINS",
            "stringValue": "{{ .contractID }}",
            "caseSensitive": true
          }
        ]
      }
    }
  ]
}
```
And this is the result of the first execution :

```
$ export MGT_ZONE_NAME=MyManagementZone1
$ export contractID=1
$ monaco deploy manifest.yaml -v
2023/03/23 11:21:28 DEBUG Writing logs to log file .logs/20230323-112128.log
2023/03/23 11:21:28 DEBUG request log not activated
2023/03/23 11:21:28 DEBUG response log not activated
2023/03/23 11:21:28 DEBUG Loading manifest "/mnt/c/Users/M4623613/Desktop/TOOLS/MobaXterm_Pro_Portable/home/git/Dynatrace/deployConfigv2/manifest.yaml". Restrictions: groups=[], environments=[]
2023/03/23 11:21:28 DEBUG Loading project `OSI` (OSI)...
2023/03/23 11:21:28 DEBUG Loading template for OSI/management-zone.json
2023/03/23 11:21:28 INFO  Projects to be deployed:
2023/03/23 11:21:28 INFO    - OSI
2023/03/23 11:21:28 INFO  Environments to deploy to:
2023/03/23 11:21:28 INFO    - cloudfactory
2023/03/23 11:21:28 INFO  Deploying configurations to environment `cloudfactory`...
2023/03/23 11:21:28 INFO        Deploying config OSI:builtin:management-zones:create-ManagementZone
2023/03/23 11:21:28 DEBUG       Upserted object create-ManagementZone (builtin:management-zones) with externalId monaco:YnVpbHRpbjptYW5hZ2VtZW50LXpvbmVzJGNyZWF0ZS1NYW5hZ2VtZW50Wm9uZQ==
2023/03/23 11:21:28 INFO  Deployment finished without errors
```
There are no problems so far. The management zone has been created correctly in Dynatrace. But when I re-execute the command by changing the environment variables with other values, the existing management zone is updated :

```
$ export MGT_ZONE_NAME=MyManagementZone2
$ export contractID=2
$ monaco deploy manifest.yaml -v
2023/03/23 11:33:13 DEBUG Writing logs to log file .logs/20230323-113313.log
2023/03/23 11:33:13 DEBUG request log not activated
2023/03/23 11:33:13 DEBUG response log not activated
2023/03/23 11:33:13 DEBUG Loading manifest "/mnt/c/Users/M4623613/Desktop/TOOLS/MobaXterm_Pro_Portable/home/git/Dynatrace/deployConfigv2/manifest-OSI.yaml". Restrictions: groups=[], environments=[]
2023/03/23 11:33:13 DEBUG Loading project `OSI` (OSI)...
2023/03/23 11:33:13 DEBUG Loading template for OSI/management-zone.json
2023/03/23 11:33:13 INFO  Projects to be deployed:
2023/03/23 11:33:13 INFO    - OSI
2023/03/23 11:33:13 INFO  Environments to deploy to:
2023/03/23 11:33:13 INFO    - cloudfactory
2023/03/23 11:33:13 INFO  Deploying configurations to environment `cloudfactory`...
2023/03/23 11:33:13 INFO        Deploying config OSI:builtin:management-zones:create-ManagementZone
2023/03/23 11:33:14 DEBUG       Upserted object create-ManagementZone (builtin:management-zones) with externalId monaco:YnVpbHRpbjptYW5hZ2VtZW50LXpvbmVzJGNyZWF0ZS1NYW5hZ2VtZW50Wm9uZQ==
2023/03/23 11:33:14 INFO  Deployment finished without errors
```
Is it the expected behaviour ? Indeed, I expected monaco to create a new MZ instead of updating the existing one because the names are different. But monaco use the same "_externalId_" to update the existing object and I don't understand why. 
Unless I am mistaken, we will able to use variable environments with "_{{ .Env.Variable }}_" in the previous releases of monaco : 

**management-zone.yaml**
```
config:
    - mgt-zone: "management-zone.json"
  
mgt-zone:
    - name: "{{.Env.MGT_ZONE_NAME}}"
```
I also tried with this syntax but without any success.
Thanks in advance for your help.

Regards,
Matthieu

**Comment by UnseenWizzard** [comment, marked as answer]
Hi @mackermann2 

yes this is expected behavior. 

Settings 2.0 objects are not identified by there name, but via the externalID you also see in the log output. 

externalID is based on the schemaId and the Monaco configID.
So even if you change the name via different environment variables, the configID will stay "create-ManagementZone" and result in the same object being identified and updated. 

You will need to be explicit, and define more than one config

---

Note: The templating tried in @mackermann2 's reply to this does not work. 

We assume that the ID in the file is unique, and there is no built-in way to define variables in it.
The general assumption here is that these files are the config as code "source of truth" for your Environment configuration - so they should be complete.

For an alternative external way to achieve the desired templating see:
https://github.com/Dynatrace/dynatrace-configuration-as-code/discussions/916#discussioncomment-5420322

**Reply by mackermann2** [reply]
Hi @UnseenWizzard 

Thanks for your reply. I understand better now. 
I confirm that I was able to create a new management zone by modifying in addition the configuration id in the YAML file : 
```
configs:
        - id: "create-ManagementZone{{.MGT_ZONE_NAME}}"
          type:
                settings:
                        schema: builtin:management-zones
                        scope: environment
          config:
                  name: 
                        type: environment
                        name: MGT_ZONE_NAME
                  parameters:
                        contractID: 
                                type: environment
                                name: contractID
                  template: management-zone.json

```
Thanks a lot for your valuable help !

**Comment by mackermann2** [comment]
I thought that it was OK but I probably forgot to delete the management zones from the Dynatrace UI and created in my previous tests...
Is it possible to variabilize the config id with a variable? Or is it necessary to have a different static YAML configuration file for each management zone to be created ? 

_Edit: or add all the new entries in the same YAML file but with a static value for config id for each MZ that I would like to create_

**Comment by mackermann2** [comment]
Hi, 
For those interested, I finally managed the modification of the configuration id with a template file :

**mgt-zone-template-yaml.txt**
```
configs:
        - id: "create-ManagementZone-##MGT_ZONE_NAME##"
          type:
                settings:
                        schema: builtin:management-zones
                        scope: environment
          config:
                  name: 
                        type: environment
                        name: MGT_ZONE_NAME
                  parameters:
                        contractID: 
                                type: environment
                                name: contractID
                  template: management-zone.json
```
where I modify the string "_##MGT_ZONE_NAME##_" from an external bash script (with sed command) with the value I want :
```
MGT_ZONE_NAME=MyMZ1
sed -e "s/##MGT_ZONE_NAME##/$MGT_ZONE_NAME/" mgt-zone-template-yaml.txt > management-zone.yaml
monaco -v deploy manifest.yaml
```
It works well for me. 
I'm putting this here. I hope it can help someone :)

**Reply by UnseenWizzard** [reply]
Thanks for adding this! 

To also reply to your question above

> Is it possible to variabilize the config id with a variable? Or is it necessary to have a different static YAML configuration file for each management zone to be created ?

No, this is not possible - we assume that the ID in the file is unique and there is no built-in way to define variables in it. 
The general assumption here is that these files are the config as code "source of truth" for your Environment configuration - so they should be complete. 

If you wish to use monaco differently you need external ways to make it work, like your sed sample :)
